pkg/router/template/util/util.go

```diff
@@ -94,3 +104,130 @@ func GenerateBackendNamePrefix(termination routev1.TLSTerminationType) string {
 
 	return prefix
 }
+
+var haproxyDurationRE = regexp.MustCompile(`^(([0-9]+)(us|ms|s|m|h|d)?)+$`)
+
+// HaproxyMaxTimeoutDuration is HaproxyMaxTimeout as a time.Duration value.
+var HaproxyMaxTimeoutDuration = func() time.Duration {
+	d, err := time.ParseDuration(HaproxyMaxTimeout)
+	if err != nil {
+		panic(err)
+	}
+	return d
+}()
+
+// OverflowError represents an overflow error from ParseHAProxyDuration.
+// OverflowError is returned if the value is greater than what time.ParseDuration
+// allows (value must be representable as uint64, e.g. approximately 2562047.79 hours
+// or 9223372036854775807 nanoseconds).
+type OverflowError struct {
+	error
+}
+
+// InvalidInputError represents an error based on invalid input to ParseHAProxyDuration
+type InvalidInputError struct {
+	error
+}
+
+// ParseHAProxyDuration is similar to time.ParseDuration, but modified with meaningful
+// error messages for invalid input, with support for decimal points removed, with
+// support for the "d" unit for days, and with "ms" as the default unit.
+// It parses a duration string and returns a suitable duration. A duration string is a
+// sequence of decimal numbers, each with optional fraction and a unit suffix,
+// such as "300ms", "1.5h" or "2h45m".  If the unit suffix is omitted, "ms" is used.
+// Valid time units are "us", "ms", "s", "m", "h", and "d".
+func ParseHAProxyDuration(s string) (time.Duration, error) {
+	orig := s
+	var d uint64
+
+	// Compare the input to the HAProxy duration regex.
+	if !haproxyDurationRE.MatchString(s) {
+		return 0, InvalidInputError{errors.New("invalid duration, no pattern match " + s)}
+	}
+	// Set the default HAProxy unit of "ms", in case a unit wasn't provided.
+	u := "ms"
```

Comment:
For correctness, `u` needs to be initialized inside the loop so that an explicit unit for one quantity doesn't determine the default unit for the subsequent quantity.  (Stylistically, `u` should also be declared inside the loop because it isn't used outside the loop.)  Since you moved the declaration and initialization outside the loop, this test case for `TestClipHAProxyTimeoutValue` now fails:
```go
		{
			value:    "24d1",
			expected: "24d1",
		},
```
The test failure indicates that `ParseHAProxyDuration` parses the value as 24 days and 1 day, which gets capped to the maximum value:
```
--- FAIL: TestClipHAProxyTimeoutValue (0.00s)
    template_helper_test.go:884: clipHAProxyTimeoutValue yielded incorrect result: expected 24d1 but got 2147483647ms
```

Having made this observation, a couple other questions come to mind that I should have raised during earlier reviews:  First, what do you think about adding the "ms" suffix when it is missing, so that the unit is obvious from reading the rendered configuration file?  So the test case would look like this:
```go
		{
			value:    "24d1",
			expected: "24d1ms",
		},
```
Second, and more importantly, should `clipHAProxyTimeoutValue`/`PArseHAProxyDuration` even allow multiple units?  `$timeSpecPattern` does not, and I believe HAProxy does not.  Even if we modify `$timeSpecPattern` to allow multiple units, we need to convert to a single unit for HAProxy, so the test case would look like this:
```go
		{
			value:    "24d1",
			expected: "2073600001ms",
		},
```
Likewise, the input value "10" would get written out in the configuration file as "10ms".

Sorry, I should have noticed this during earlier reviews.

Reply:
Thanks @Miciah.  I've removed the ability to have multiple units as part of the string, so 1d1m is not allowed any more.  I changed the regex so multiples are not allowed, and updated the comments and unit tests.  I also moved the initialization of "u" into the loop.  PTAL when you get a chance and let me know if this is now acceptable.